### ISSUE TITLE
Inserting spaces inside the code element does not work

### ISSUE THREAD

**Issue by pomek** [ckeditor/ckeditor5-engine#1116]
_All scenarios require the `Code` feature which will be introduced in https://github.com/ckeditor/ckeditor5-basic-styles/pull/55._

The following scenarios shouldn't crash the editor:

#### 1. Insert a space at the end of `</code>` element.

Make an editor with the following content:

```html
<p><i>This</i> <code>is an[]</code> <strong>editor</strong> <u>instance</u>.</p>
```

Then insert a space. The editor data should looks like:

```html
<p><i>This</i> <code>is an []</code> <strong>editor</strong> <u>instance</u>.</p>
```

Now we can notice the first problem – one of the space is not visible, two spaces are next to each other (in DOM). 

![image](https://user-images.githubusercontent.com/2270764/29915658-c3a7e97e-8e3c-11e7-9e6b-4cddc6c96741.png)

After inserting another space, the editor will blow up. However, the editor still returns a proper HTML:

```html
<p><i>This</i> <code>is an  []</code> <strong>editor</strong> <u>instance</u>.</p>
```

Editor crashes on the code related to mutations.

```
Uncaught TypeError: Cannot read property 'index' of undefined
    at MutationHandler._handleTextNodeInsertion (input.js:298)
    at MutationHandler.handle (input.js:152)
    at Input._handleMutations (input.js:105)
    at Document.listenTo (input.js:50)
    at Document.fire (emittermixin.js:281)
    at MutationObserver._onMutations (mutationobserver.js:248)
```

[`const change = getSingleTextNodeChange( mutation );`](https://github.com/ckeditor/ckeditor5-typing/blob/cef7405b601cc8027d59e2900f45ad88cb4d2c16/src/input.js#L293) –  this expression for the specific case is `undefined`. When I added `if ( !changes ) { return; }`, the error does not occur but another spaces are not inserted.

Temporary solution (thanks to @szymonkups) is replacing this line – https://github.com/ckeditor/ckeditor5-engine/blob/a424116abc84287d49343898d84ac471c4c21932/src/view/renderer.js#L472 with the following code:

```js
let expectedDomChildren = Array.from( domConverter.viewChildrenToDom( viewElement, domDocument, { bind: true } ) );

if ( viewElement.name == 'code' ) {
    expectedDomChildren = expectedDomChildren.map( item => {
        if ( item.nodeType === global.window.Node.TEXT_NODE ) {                
            item.data = item.data.replace( / /g, '\u00A0' );
        }

        return item;
    } );
}
```

but it is an ugly hack.

I wrote the test which confirms that model and view have proper values – https://gist.github.com/pomek/912a7c9b391bb1442b0d825b13d2b17e.

It looks like something is invalid with the `code` element.

#### 2. Insert space as a code in empty editor

1. Clear the editor.
2. Use `code` styles.
3. Press space.

The data stored in the model are correct:

```html
<p><code> []</code></p>
```
Rendered DOM also is ok but the editor does not look ok. You can compare the editor views' by inserting an empty bolded space instead of the code.

![image](https://user-images.githubusercontent.com/2270764/29915673-d44a3296-8e3c-11e7-9d44-5c99e5e07214.png)

![image](https://user-images.githubusercontent.com/2270764/29915450-3bb449f4-8e3c-11e7-82bd-d4baba6449a4.png)

4. After inserting another space:
  * Model:

  ```html
  <code> </code><p><code> </code></p>
  ```

  * Rendered DOM:

  ```html
  <code> </code><p><code> </code></p>
  ```
  
5. Inserting another space will blow up the editor:

```
Uncaught TypeError: Cannot read property 'index' of undefined
```

After pasting temporary solution in `engine/view/renderer~Renderer#_updateChildren()`, the error will occur once and the spaces will be inserted.

**Comment by scofalik** [comment]
I did changes recently in handling multiple spaces so this may be related with the code I was workin with. I'll look at it. For sure we cannot apply a hack like suggested.

**Comment by Reinmar** [comment]
I saw recently, when styling the docs, that whitespaces at the boundaries of `<code>` are not preserved by the browser. So this may be just an issue with this. I had to use a styling like this to preserve them:

```
	code:not( .hljs ) {
		// It's a darker shade of u-color( 'background-hue' ) so it looks good in info-boxes.
		// See #230.
		background: rgba( 202, 205, 207, 0.3 );
		padding: u-spacing( -4 ) u-spacing( -3 );
		border-radius: 3px;
	}

	// See #223.
	code:not( .hljs ):after {
		// To reduce the width of that space to 0px.
		letter-spacing: -1em;
		content: "\00a0";
	}

	// See #223.
	code:not( .hljs ):before {
		// To reduce the width of that space to 0px.
		letter-spacing: -1em;
		content: "\00a0";
	}
```

**Comment by Reinmar** [comment]
If that's the case, then the engine does not need to do anything.

**Comment by pomek** [comment]
I modified the CSS a little:

```scss
code:not( .hljs ) {
	// It's a darker shade of u-color( 'background-hue' ) so it looks good in info-boxes.
	// See #230.
	background: rgba( 202, 205, 207, 0.3 );
	padding: u-spacing( -4 ) u-spacing( -3 );
	border-radius: 3px;

	// See #223.
	&:before,
	&:after {
		// To reduce the width of that space to 0px.
		letter-spacing: -1em;
		content: "\00a0";
	}

	& > br[data-cke-filler] {
		display: none;
	}
}
```

and now **the editor does not blow up**!

But, we cannot add more than one space before the `</code>` (but I guess it isn't a big problem – https://github.com/ckeditor/ckeditor5-basic-styles/issues/52#issuecomment-323987258).

**Comment by pomek** [comment]
I found another problem – you can type two spaces before closing the `</code>` tag but only one space will be rendered. 

```html
<p><i>This</i> <code>is an  []</code> <strong>editor</strong> <u>instance</u>.</p>
```

![image](https://user-images.githubusercontent.com/2270764/29919165-7bd79cd2-8e48-11e7-8a4d-0a2bc1217535.png)

When you press <kbd>Backspace</kbd>, you will get:

```html
<p><i>This</i> <code>is an []</code> <strong>editor</strong> <u>instance</u>.</p>
```

But the rendered view will not change. 

![image](https://user-images.githubusercontent.com/2270764/29919183-8dc90520-8e48-11e7-9d12-e84e36b70f92.png)

If you deselect the editor, then focus once again at the end of the code, then press the <kbd>Backspace</kbd>, you will get:

```html
<p><i>This</i> <code>is an[] </code> <strong>editor</strong> <u>instance</u>.</p>
```

As a summary: the editor will work but the spaces are incorrect rendered.

**Comment by Reinmar** [comment]
**tl;dr here is that after introducing proper styles typing more than one space in a row anywhere in `<code>` doesn't work.**

**Comment by Reinmar** [comment]
Additionally, an error is thrown in Safari when typing spaces at the end of `<code>`.

**Comment by Reinmar** [comment]
We'll approach this from the other side: https://github.com/ckeditor/ckeditor5-engine/issues/1126
